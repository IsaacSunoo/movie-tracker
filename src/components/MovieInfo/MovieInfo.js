import React, { Component } from 'react';
import { connect } from 'react-redux';
import { toggleFavorite } from '../../actions/actions';
import { Link } from 'react-router-dom';
import { postFetch, deleteFetch } from '../../api';

class MovieInfo extends Component {

  handleFavorite = (movieId) => {
    const { favorites } = this.props;
    if (!favorites.includes(movieId)) {
      this.addFavorite();
    } else {
      this.deleteFavorite(movieId);
    }
  }

  addFavorite = async () => {
    const url = 'users/favorites/new'
    const { id, title, poster_path, release_date, vote_average, overview, loginUser, toggleFavorite } = this.props;
    console.log('movie info props: ', this.props)
    const favInfo = { movie_id: id, user_id: loginUser.id, title, poster_path, release_date, vote_average, overview };
    await postFetch(url, 'POST', favInfo);
    toggleFavorite(id);
  }

  deleteFavorite = async (movieId) => {
    const { id, toggleFavorite, loginUser } = this.props;
    const url = `users/${loginUser.id}/favorites/${movieId}`;
    await deleteFetch(url);
    toggleFavorite(id);
  }

  render() {
    const { id, title, overview, poster_path, release_date, vote_average, vote_count, backdrop_path, favorites } = this.props;
    const imageSrc = {backgroundImage: `url(http://image.tmdb.org/t/p/w342//${poster_path})` };
    const background = { backgroundImage: `url(http://image.tmdb.org/t/p/original${backdrop_path})` };

    const star = (
      <div className="star" onClick={() => { this.handleFavorite(id) }}>
        {!favorites.includes(id) && <i className="fas fa-star" />}
        {favorites.includes(id) && <i className="fas fa-star active-favorite" />}
      </div>
    );

    return (
      <div className='movie-info'>
        <div className='movie-image' style={imageSrc}></div>
        <article className='info-section' style={background}>
          <div className='overlay'>
            <div className='movie-title'>
              {star}
              <h2>{title}</h2>
              <p className='release-date'>Release Date: {release_date}</p>
            </div>
            <div className='review'>
              <p>Rated <span>{vote_average}</span>/10</p>
              <p className='vote-count'>({vote_count} viewer votes)</p>
            </div>
            <div className='movie-description'>
              <h4>Description</h4>
              <p>{overview}</p>
            </div>
            <Link to='/'>
              <p className='return'>Return to all movies</p>
            </Link>
          </div>
        </article>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  loginUser: state.loginUser,
  favorites: state.favorites
});

const mapDispatchToProps = dispatch => ({
  toggleFavorite: id => dispatch(toggleFavorite(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieInfo);