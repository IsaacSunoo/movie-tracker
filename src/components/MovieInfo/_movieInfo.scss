.movie-info {
  border: solid $ground 3px;
  border-radius: 15px;
  color: $smoke;
  display: grid;
  grid-template-columns: 40% 60%;
  height: 500px;
  left: 5%;
  margin-top: 20px;
  overflow: hidden;
  position: relative;
  width: 90%;
}

.movie-image {
  background-size: cover;
  background-position: center;
  border-right: solid $ground 3px;
  height: 100%;
  width: 100%;
  overflow: hidden;
}

.overlay {
  background: rgba(255,255,255,.7);
  height: 100%;
  left: 0;
  // position: absolute;
}

.info-section {
  background-size: cover;
  background-position: center;
  height: 100%;
  text-align: center;
  position: relative;
} 

.movie-title {
  h2 {
    font-size: 260%;
    padding: 20px;
  }
}

.release-date {
  font-size: 140%;
  margin-bottom: 20px;
}

.review {
   font-size: 140%;
   span {
    color: red;
   }
   .vote-count {
    font-size: 75%
   } 
}

.movie-description {
  padding: 20px;
  h4 {
    font-size: 160%;
    margin-bottom: 20px;
    text-decoration: underline;
  }
  p {
    font-size: 18px;
  }
}