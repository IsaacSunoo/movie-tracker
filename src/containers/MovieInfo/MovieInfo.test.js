import React from 'react';
import { MovieInfo, mapStateToProps, mapDispatchToProps } from './MovieInfo';
import { shallow } from 'enzyme';

describe('MovieInfo', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <MovieInfo />
    )
  });
  
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  
  describe('mapStateToProps', () => {
    it('should return an object with currentuser and favorites', () => {
      const mockState = {
        loginUser: {name: 'Isaac'},
        favorites: [123, 444],
        movies: [{title: 'Hey'}, {title: 'anotherMovie'}]
      }
      const expected = {
        loginUser: {name: 'Isaac'},
        favorites: [123, 444]
      }
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected)
    });
  });
  describe('mapDispatchToProps', () => {
    it('should call dispatch when using toggleFavorite', () => {
      const mockDispatch = jest.fn();
      const action = actions.toggleFavorite({name: 'Mason'})
      const mappedDispatch = mapDispatchToProps(mockDispatch)
      mappedDispatch.toggleFavorite({name: 'Mason'})
      expect(mockDispatch).toHaveBeenCalledWith(action)
    });
  });
  describe('handleFavorite', () => {
    
  });

  describe('addFavorite', () => {

  });

  describe('deleteFavorite', () => {

  });
});