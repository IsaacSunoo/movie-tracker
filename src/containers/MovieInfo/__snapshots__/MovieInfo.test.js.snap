// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieInfo should match snapshot 1`] = `undefined`;
